modin/core/dataframe/base/partitioning/axis_partition.py

```diff
@@ -0,0 +1,161 @@
+            Another `BaseDataframeAxisPartition` object to be applied
+            to func. This is for operations that are between two data sets.
+        maintain_partitioning : bool, default: True
+            Whether to keep the partitioning in the same
+            orientation as it was previously or not. This is important because we may be
+            operating on an individual axis partition and not touching the rest.
+            In this case, we have to return the partitioning to its previous
+            orientation (the lengths will remain the same). This is ignored between
+            two axis partitions.
+        **kwargs : dict
+            Additional keywords arguments to be passed in `func`.
+
+        Returns
+        -------
+        list
+            A list of `BaseDataframePartition` objects.
+
+        Notes
+        -----
+        The procedures that invoke this method assume full axis
+        knowledge. Implement this method accordingly.
+
+        You must return a list of `BaseDataframePartition` objects from this method.
+        """
+        pass
+
+    def shuffle(self, func, lengths, **kwargs):
+        """
+        Shuffle the order of the data in this axis partition based on the `lengths`.
+
+        Parameters
+        ----------
+        func : callable
+            The function to apply before splitting.
+        lengths : list
+            The list of partition lengths to split the result into.
+        **kwargs : dict
+            Additional keywords arguments to be passed in `func`.
+
+        Returns
+        -------
+        list
+            A list of `BaseDataframePartition` objects split by `lengths`.
+        """
+        pass
+
+    # Child classes must have these in order to correctly subclass.
+    instance_type = None
+    partition_type = None
+
+    def _wrap_partitions(self, partitions):
+        """
+        Wrap remote partition objects with `BaseDataframePartition` class.
+
+        Parameters
+        ----------
+        partitions : list
+            List of remotes partition objects to be wrapped with `BaseDataframePartition` class.
+
+        Returns
+        -------
+        list
+            List of wrapped remote partition objects.
+        """
+        return [self.partition_type(obj) for obj in partitions]
+
+    def force_materialization(self, get_ip=False):
+        """
+        Materialize axis partitions into a single partition.
+
+        Parameters
+        ----------
+        get_ip : bool, default: False
+            Whether to get node ip address to a single partition or not.
+
+        Returns
+        -------
+        BaseDataframeAxisPartition
+            An axis partition containing only a single materialized partition.
+        """
+        materialized = self.apply(
+            lambda x: x, num_splits=1, maintain_partitioning=False
+        )
+        return type(self)(materialized, get_ip=get_ip)
```

Comment:
why do we have an implementation in abstract base?..

Reply:
It's legacy but yes we want to keep abstract base classes with methods declaration only. Should probably be fixed as part of a separate issue.

Reply:
Could you please make a separate issue then?

Reply:
Sure, #3635